### ISSUE TITLE
Exact solver

### ISSUE THREAD

**Issue by ericphanson** [evanfields/TravelingSalesmanHeuristics.jl#18]
I've been (recreationally) learning about some TSP techniques and implemented an exact (slow) method here, with a specialized implementation for the symmetric case: https://github.com/ericphanson/TSP_MIP.jl. I thought it could be useful for your package for generating test cases for benchmarking heuristic performance, and also the exact solver could benefit from using the heuristic solutions as a warmstart (but that's not implemented yet in the GPLK wrapper, tracked here I believe: https://github.com/JuliaOpt/GLPK.jl/issues/82). I actually changed the API a bit to match yours so it should be easy to use together; I saw the [discourse post](https://discourse.julialang.org/t/type-hierarchy-for-tsp-instances/23123) just now and could use such an API for the exact solver too (I decided to open an issue to not derail that thread). I'm also happy to contribute it to your package in a pull request if that's something you'd be interested in.

**Comment by evanfields** [comment]
Your package looks great!

I've thought before about adding an exact solver to this package, either as normal dependency or an optional dependency via Requires.jl. I'm definitely persuadable and eager to hear your thoughts, but at present it seems like:
- A required dependency, especially one that in turn depends on GLPK, etc., would make us no longer be lightweight. (Currently no dependencies!)
- An optional dependency still requires Requires.jl, and doesn't help the user that much relative to just pointing out the existence of the exact package.
- With the possible exception of MIP warmstarts, there's not much interdependence between a TSP heuristic and exact package.

If you're planning to add TSP_MIP to the general registry and maintain it, I'd be delighted to link prominently to it in TravelingSalesmanHeuristic's readme and documentation. 

The point about using the exact method for benchmarks and test cases is well taken. `solve_tsp` has a naive polyalgorithm that could definitely use some love, and an easy exact method would be a real help there.

All that said, I'm eager to hear your thoughts and/or entertain any PRs. I'm not particularly experienced with open source and am sure I have a lot to learn!

**Comment by ericphanson** [comment]
Thanks, your package looks great too!

The dependency issue makes a lot of sense. The approach JuMP takes is to not depend on any solver and then have the user depend on both JuMP and a solver. I could take that approach here which would remove the dependency on GPLK, which maybe is a good idea for me to do regardless; I decided to depend on the solver just so it would be as easy as possible to get started. But either way, maybe it makes more sense to have the exact algorithms in a separate package since then you don’t have to depend on JuMP either or e.g. UnicodePlots. (Since the exact code is slow anyway, I think it’s nice to do things like plotting the progress so you can see what’s going on while you wait and to understand the algorithm better, but I can definitely understand not wanting to add those dependencies.) 

I can register my code (maybe as TravelingSalesmanExact or something like that), and I can just make sure to implement the same API so it’s easy to benchmark and make test cases. I’m not sure I’ll have a ton of time to work on the package but it should be easy for me to maintain it at least (it uses JuMP 19 which is after their big shift to MathOptInterface so I think it’ll be fairly stable). I might try to add other exact methods as well, but I’m not sure other ways will be much better compared to just waiting for warmstarts and using heuristics to speed up the approach I’m already using. 

And to benchmark for example, we don’t really need one package to be a dependency of the other; just a notebook or something that uses both packages as done in the DiffEq project: https://github.com/JuliaDiffEq/DiffEqBenchmarks.jl. I think the DiffEq project is a good one to look to since my understanding is that they use their benchmarks to tune their polyalgorithm which decides what differential equation solver to use if the user doesn’t specify one. I really don’t know anything about making polyalgorithms but it seems very interesting and I’d be happy to try to collaborate on using benchmarks to tune the one in your package.

(I am fairly new to open source as well!)

**Comment by ericphanson** [comment]
I went ahead and made <https://github.com/ericphanson/TravelingSalesmanBenchmarks.jl> with some simple benchmarks (can be viewed here: https://ericphanson.github.io/TravelingSalesmanBenchmarks.jl/), based on `DiffEqBenchmarks`. Let me know your thoughts! I'm happy to add you as a collaborator if you approve of the idea and would like commit access to add more benchmarks etc.

I also changed my package's name to `TravelingSalesmanExact` and removed the `GLPK` dependency by requiring the user to instead pass an optimizer. For me, `Mosek` runs much faster and I have a license installed anyway, so it's more convenient to not require `GLPK`. I'll try and register it soon.
